### ISSUE TITLE
[BUG] - [D&D5e] - Merchant Filters only work for strings

### ISSUE THREAD

**Issue by MrFlaig** [fantasycalendar/FoundryVTT-ItemPiles#659]
**I confirm the following:**
- [Yes] I am on the latest version of Item Piles
- [Yes] I have tried to clear my cache (Ctrl / Cmd + F5)
- [Yes] I have checked the console for errors (F12 to open, Console tab)
- [Yes] I have tried to reset Item Piles' system settings
- [Yes] I deactivated any others modules

**Describe the bug**
If I try to use arrays or integers as filters in the merchant settings no items show up in the "Sell Items" interface, even if the actor cleary has items with the used properties

*To Reproduce**

1. Setup a Merchant
2. Use any filters that are not strings

**Expected behavior**
Item's are shown for the selected filters.

**Setup**
- Foundry Version: [12.331]
- System Version: [DND5e 3.3.1]
- Item Piles Version: [3.1.5]
- Browser & version [Google Chrome 128.0.6613.138]

**Active modules**
- Item Piles
- libwrapper
- socketlib
- Tested it with and without any other modules

**Comment by MrFlaig** [comment]
I let "something" take a look at the two functions in question and it gave me this. Maybe it helps. It's working for me at least, while I tested it with arrays, but don't know if anything else needs to be done.

```
function isItemValidBasedOnProperties(targetActor, item) {
  const pileItemRequiredProperties = getActorRequiredItemProperties(targetActor);
  const itemData = item instanceof Item ? item.toObject() : item;

  function isMatch(attributeValue, filterValue) {
    if (Array.isArray(attributeValue)) {
      return attributeValue.some(value => isMatch(value, filterValue));
    }
    if (typeof attributeValue === 'number' && typeof filterValue === 'number') {
      return Math.abs(attributeValue - filterValue) < Number.EPSILON;
    }
    return attributeValue === filterValue;
  }

  for (const filter of pileItemRequiredProperties) {
    if (!foundry.utils.hasProperty(itemData, filter.path)) {
      return false;
    }
    
    const attributeValue = foundry.utils.getProperty(itemData, filter.path);
    const filterValues = Array.isArray(filter.filters) ? filter.filters : Array.from(filter.filters);
    
    const matchFound = filterValues.some(filterValue => isMatch(attributeValue, filterValue));
    
    if (!matchFound) {
      return false;
    }
  }
  
  return true;
}
```
```
function isItemInvalid(targetActor, item, itemFilters = false) {
  const pileItemFilters = itemFilters ? itemFilters : getActorItemFilters(targetActor);
  const itemData = item instanceof Item ? item.toObject() : item;

  function isMatch(attributeValue, filterValue) {
    if (Array.isArray(attributeValue)) {
      return attributeValue.some(value => isMatch(value, filterValue));
    }
    if (typeof attributeValue === 'number' && typeof filterValue === 'number') {
      return Math.abs(attributeValue - filterValue) < Number.EPSILON;
    }
    return attributeValue === filterValue;
  }

  for (const filter of pileItemFilters) {
    if (!foundry.utils.hasProperty(itemData, filter.path)) {
      continue;
    }
    const attributeValue = foundry.utils.getProperty(itemData, filter.path);
    const filterValues = Array.from(filter.filters);
    
    for (const filterValue of filterValues) {
      if (isMatch(attributeValue, filterValue)) {
        return filterValue; // Return the matching filter value
      }
    }
  }
  
  return false;
}
```

**Comment by Haxxer** [comment]
Nice idea, but I would not put a function in a function, as `isItemInvalid` is used extensively. Let's declare it outside of the function, so that it is not re-created for every single call of the function.

**Comment by MrFlaig** [comment]
So it should look something like this? It's at least working for me.

```
// Helper function used by both isItemInvalid and isItemValidBasedOnProperties
function isMatch(attributeValue, filterValue) {
  if (Array.isArray(attributeValue)) {
    return attributeValue.some(value => isMatch(value, filterValue));
  }
  if (typeof attributeValue === 'number' && typeof filterValue === 'number') {
    return Math.abs(attributeValue - filterValue) < Number.EPSILON;
  }
  return attributeValue === filterValue;
}

function isItemInvalid(targetActor, item, itemFilters = false) {
  const pileItemFilters = itemFilters ? itemFilters : getActorItemFilters(targetActor);
  const itemData = item instanceof Item ? item.toObject() : item;

  for (const filter of pileItemFilters) {
    if (!foundry.utils.hasProperty(itemData, filter.path)) {
      continue;
    }
    const attributeValue = foundry.utils.getProperty(itemData, filter.path);
    const filterValues = Array.from(filter.filters);
    
    for (const filterValue of filterValues) {
      if (isMatch(attributeValue, filterValue)) {
        return filterValue; // Return the matching filter value
      }
    }
  }
  
  return false;
}

function isItemValidBasedOnProperties(targetActor, item) {
  const pileItemRequiredProperties = getActorRequiredItemProperties(targetActor);
  const itemData = item instanceof Item ? item.toObject() : item;

  for (const filter of pileItemRequiredProperties) {
    if (!foundry.utils.hasProperty(itemData, filter.path)) {
      return false;
    }
    
    const attributeValue = foundry.utils.getProperty(itemData, filter.path);
    const filterValues = Array.isArray(filter.filters) ? filter.filters : Array.from(filter.filters);
    
    const matchFound = filterValues.some(filterValue => isMatch(attributeValue, filterValue));
    
    if (!matchFound) {
      return false;
    }
  }
  
  return true;
}
```

**Comment by Haxxer** [comment]
Fixed in 3.1.6, thanks for your code suggestion
